illumos-utils/src/zfs.rs

```diff
@@ -203,14 +207,84 @@ pub struct EncryptionDetails {
 
 #[derive(Debug, Default)]
 pub struct SizeDetails {
-    pub quota: Option<usize>,
-    pub reservation: Option<usize>,
+    pub quota: Option<ByteCount>,
+    pub reservation: Option<ByteCount>,
     pub compression: CompressionAlgorithm,
 }
 
+#[derive(Debug)]
+pub struct DatasetProperties {
+    /// The Uuid of the dataset
+    pub id: Option<DatasetUuid>,
+    /// The full name of the dataset.
+    pub name: String,
+    /// Remaining space in the dataset and descendents.
+    pub avail: ByteCount,
+    /// Space used by dataset and descendents.
+    pub used: ByteCount,
+    /// Maximum space usable by dataset and descendents.
+    pub quota: Option<ByteCount>,
+    /// Minimum space guaranteed to dataset and descendents.
+    pub reservation: Option<ByteCount>,
+    /// The compression algorithm used for this dataset.
+    pub compression: String,
+}
+
+impl FromStr for DatasetProperties {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut iter = s.split_whitespace();
+
+        let id = match iter.next().context("Missing UUID")? {
+            "-" => None,
+            anything_else => Some(anything_else.parse::<DatasetUuid>()?),
+        };
+
+        let name = iter.next().context("Missing 'name'")?.to_string();
+        let avail = iter
+            .next()
+            .context("Missing 'avail'")?
+            .parse::<u64>()?
+            .try_into()?;
+        let used = iter
+            .next()
+            .context("Missing 'used'")?
+            .parse::<u64>()?
+            .try_into()?;
+        let quota =
+            match iter.next().context("Missing 'quota'")?.parse::<u64>()? {
+                0 => None,
+                q => Some(q.try_into()?),
+            };
+        let reservation = match iter
+            .next()
+            .context("Missing 'reservation'")?
+            .parse::<u64>()?
+        {
+            0 => None,
+            r => Some(r.try_into()?),
+        };
+        let compression =
+            iter.next().context("Missing 'compression'")?.to_string();
```

Comment:
These errors would ideally return the input that failed.

Reply:
I pulled this parsing implementation into a helper; I'm now always emitting the input string in the error message.